prbt_hardware_support/src/system_info.cpp

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2019 Pilz GmbH & Co. KG
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <prbt_hardware_support/system_info.h>
+
+#include <XmlRpcValue.h>
+#include <XmlRpcException.h>
+#include <sensor_msgs/JointState.h>
+
+#include <canopen_chain_node/GetObject.h>
+#include <prbt_hardware_support/system_info_exception.h>
+
+namespace prbt_hardware_support
+{
+
+static const std::string CANOPEN_GETOBJECT_SERVICE_NAME{"/prbt/driver/get_object"};
+static const std::string CANOPEN_NODES_PARAMETER_NAME{"/prbt/driver/nodes"};
+static const std::string JOINT_STATE_TOPIC {"/joint_states"};
+static constexpr double WAIT_FOR_SERVICE_TIMEOUT_S{3.0};
+static constexpr double JOINT_STATES_TOPIC_TIMEOUT{3.0};
```

Comment:
Yes. I would agree. However, then the tests will take forever ;)

Reply:
As discussed in #81 it is always a bad idea, to abort node startup with such a fixed timeout. Please refactor as in #122.

Reply:
> As discussed in #81 it is always a bad idea, to abort node startup with such a fixed timeout. Please refactor as in #122.
I do not want to disagree but in my opinion #122 is even worse especially the lines [here](https://github.com/PilzDE/pilz_robots/blob/de39cc0f9bedeefa85efe822ef25daece1b59e5e/prbt_hardware_support/src/sto_modbus_adapter.cpp#L45) because that way you might end up with a system which will never reach a proper state instead of failing.

Reply:
Yep. As discussed there: RQT and other ros-tools use such a blocking approach. They print a message "waiting for topic ..." and block. That's common in "the ROS world" and we should imho use that design as well.
It has the advantage that the user can split launch files and start different nodes at different times (other nodes waiting). An exception doesn't allow such a use (even worse in the case of the modbus node, that is marked as `required` and thus kills all other nodes as well).
Of course you have to check for `ros::ok()` to allow aborting the loop. Please see also the example in [ros wiki](http://wiki.ros.org/roscpp/Overview/Logging).

Reply:
> As discussed in #81 it is always a bad idea, to abort node startup with such a fixed timeout. Please refactor as in #122.

Done.